Field transformers

cvalarida opened the pull request:
## Description
This is a prototype for transforming entity bundles (what we've been calling "content models") without writing a transformer for each bundle.

The approach here is to transform each field individually by mapping the schema for the input to the schema for the output. We can write a single function for each mapping and as we add new bundles and change the existing content model, as long as the schemas accurately represent the data and we have field transformers for each input -> output combination, the changes will be accounted for automatically in the transformer step.

## Testing done
I wrote an escape hatch in `assembleEntityTree` to bypass the current process for `node-vba_facility`, a new bundle which we had no schemas or transformer for. I plugged in `console.log`s as I was working to verify each step, but the real test is done when you run `yarn build:content --use-cms-export`. See the screenshot below for a snippet of the result. Note that the post-transformation validation is running, so that it doesn't throw any errors indicates the process is working as intended.

## Screenshots
![image](https://user-images.githubusercontent.com/12970166/85336349-4fc93500-b493-11ea-9fae-837679bbc8f1.png)


## Acceptance criteria
- [x] The `node-vba_facility` transformation occurs without a transformer function written specifically for the bundle

Commit:
Start work on field transformers

Commit:
Bypass assembleEntityTree for node-vba_facility

Commit:
Add jsdoc param

Commit:
Deliberately call out the prototype escape hatch

Commit:
Fill in the field transformer function

Commit:
Add a helper for entity references

Commit:
Only add schemas (not helper functions) to AJV

Commit:
Add node-vba_facility schemas

Commit:
Add TODO

Commit:
Move fieldTransformers to a new module

Commit:
Set the contentModelType so we can validate the output

Commit:
TODO to register() new transformers

bkjohnson commented on src/site/stages/build/process-cms-exports/index.js:
```diff
@@ -164,6 +166,25 @@ const entityAssemblerFactory = contentDir => {
    */
   const assembleEntityTree = (entity, ancestors = [], parentFieldName = '') => {
+    // Use the automatic transformer process for certain content models
+    // This is a test. Eventually,
```
I'm guessing that eventually the `transformFields` will be the default and only way that we do this?

bkjohnson commented on src/site/stages/build/process-cms-exports/field-transformers/index.js:
```diff
@@ -0,0 +1,54 @@
+  ],
+]);
+
+module.exports = { fieldTransformers, normalize, serialize };
```
I might have overlooked something, but we might not need to export `normalize`.

bkjohnson commented on src/site/stages/build/process-cms-exports/field-transformers/index.js:
```diff
@@ -0,0 +1,54 @@
+ */
+const serialize = thing => JSON.stringify(normalize(thing), null, 2);
+
+/**
+ * Maps a transformer to an input and output schema for an individual field.
+ *
+ * TODO: Make this less annoying for enums. This may mean removing enums
+ * recursively in normalize(), but that seems maybe too specific? Not sure what
+ * else we may need to do. Probably reasonable for a prototype, though.
+ */
+const fieldTransformers = new Map([
+  [
+    // Input schema: Data from Drupal looks like this
+    serialize({ $ref: 'GenericNestedString' }),
+    new Map([
+      // Output schema: Data in the templates look like this
+      [serialize({ type: 'string' }), getDrupalValue],
+    ]),
+  ],
+]);
```
It isn't immediately obvious why we have a map within a map.  Can that be made clearer? I get the idea of "if we have a `GenericNestedString` then we want to use `getDrupalValue` to transform it"

bkjohnson commented on src/site/stages/build/process-cms-exports/transform-fields.js:
```diff
@@ -0,0 +1,148 @@
+ *                 specifying only the fields the parent needs.
+ */
+function transformFields(entity, outputSchemaFromParent) {
+  // console.log(JSON.stringify(entity, null, 2));
+  // throw new Error('stopping');
```
Are these useful?

bkjohnson commented on src/site/stages/build/process-cms-exports/transform-fields.js:
```diff
@@ -0,0 +1,148 @@
+        return result;
+      }
+
+      // Find the specific field transformer
+      const transformField = inputTransformerMap.get(
+        serialize(outputSchema.properties[outputKey]),
+      );
```
I don't quite understand the situation where that inner `Map` has multiple mappings that we care about.

bkjohnson reviewed (commented):
The main thing that I'm having a hard time wrapping my head around is `fieldTransformers`.  It seems like it is set up to allow a one-to-many mapping of input to output schemas, and I don't understand that.

Commit:
Use a registry for mapping transformers to schemas

Commit:
Automagic field transformer registration

Commit:
Register more schema combinations for getDrupalValue

Commit:
Use the field transformers on everything

Commit:
Log entity ID for help debugging

Commit:
Track output schema warnings better

Commit:
Extract function: expandEntityReferenceArray

Commit:
If there's no data, call it null

Commit:
Clean up some logging

Commit:
Update schema for use in the field transformers

Commit:
Facility hours transformer

Commit:
Epoch time transformer

Commit:
Address transformer

Commit:
Add inputKey exceptions

Because it's not always snake_case

Commit:
Extra logging

Commit:
Add epoch time transformer

Commit:
Add wysiwyg string tranformer

Commit:
Add metatags transformer

Commit:
Add URL transformer

Commit:
Add published state transformer

Commit:
Lighlty modify some schema changes

Commit:
Remove unused code

Commit:
Add some properties to every entity

contentModelType, baseType, entityBundle

Commit:
Map another schema pair to get-wysiwyg-string

Commit:
Moar TODOz!!

Commit:
Expand some schemas

Commit:
Pass the items schema when expanding entity references

Ran into a nasty bug where this created a 2D array because both
expandEntityReferenceArray and transformFields were returning arrays.
This was happening becuase I was passing the schema for the array to
transformFields from expandEntityReferenceArray, but I wanted to pass
the schema for each item when I'm iterating over each item. Because that
makes sense.

There are some times that the output schema doesn't specify an array,
though, so we still want to pass the outputFieldSchema as-is (not
.items). So I had to handle both.

Commit:
Extract function: transformObject

Commit:
Be less verbose in the logging

Commit:
JSDocs

Commit:
Merge branch 'master' into field-transformers-cv

Commit:
Don't read a node if it's not published

Also, use entity.status[0].value to determine if an entity is published.

Commit:
Remove common schemas for globbing entities

Commit:
Remove $ref from transformed schemas

Commit:
Use $contentModelType to find the anyOf / allOf schema

Commit:
Always $expand 'entity' properties
